posthog/session_recordings/queries/session_recording_list_from_query.py

```diff
@@ -744,17 +797,46 @@ def _where_predicates(self) -> ast.Expr:
     def _having_predicates(self) -> ast.Expr:
         (_, event_names) = self._event_predicates
 
+        exprs: list[ast.Expr] = []
         if event_names:
-            return ast.Call(
-                name="hasAll" if self.property_operand == PropertyOperatorType.AND else "hasAny",
-                args=[
-                    ast.Call(name="groupUniqArray", args=[ast.Field(chain=["event"])]),
-                    # KLUDGE: sorting only so that snapshot tests are consistent
-                    ast.Constant(value=sorted(event_names)),
-                ],
+            exprs.append(
+                ast.Call(
+                    name="hasAll" if self.property_operand == PropertyOperatorType.AND else "hasAny",
+                    args=[
+                        ast.Call(name="groupUniqArray", args=[ast.Field(chain=["event"])]),
+                        # KLUDGE: sorting only so that snapshot tests are consistent
+                        ast.Constant(value=sorted(event_names)),
+                    ],
+                )
             )
 
-        return ast.Constant(value=True)
+        if self.event_properties:
+            # when we're saying property is not set then we have to check it is not set on every event
+            # e.g. countIf(JSONHas(events.properties, '$feature/target-flag')) = 0
+            for prop in self.event_properties:
```

Comment:
this is the main fix so far...
when looking at a negative property like "not in" we check instead that the count of the positive equivalent is zero across the session

so we're changing the query from

> show me sessions where any event matches the negative filter

to 

> show me sessions where every event matches the negative filter